frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3566,29 +3609,21 @@ gatherAndFilterCandidates(Context* context,
 
   // If no candidates have been found, consider last resort candidates from
   // innermost to outermost.
-  for (size_t i = 0; i < lrcGroups.size(); i++) {
-    if (candidates.empty()) {
-      candidates = lrcGroups[i];
-      // Bump firstPoiCandidate index after non-POI last resorts
-      if (i == 0) {
-        firstPoiCandidate = candidates.size();
-      }
-    } else {
-      break;
-    }
+  if (candidates.empty()) {
+    candidates = lrcGroups.nonPoiCandidates();
+    firstPoiCandidate = candidates.size();
   }
-  // If still no candidates, repeat for last resort candidates from forwarding.
-  for (size_t i = 0; i < forwardingLrcGroups.size(); i++) {
-    if (candidates.empty()) {
-      candidates = forwardingLrcGroups[i];
-      forwardingInfo = forwardingLrcForwardingTo[i];
-      // Bump firstPoiCandidate index after non-POI last resorts
-      if (i == 0) {
-        firstPoiCandidate = candidates.size();
-      }
-    } else {
-      break;
-    }
+  for (size_t i = 0; candidates.empty() && i < lrcGroups.numPoiGroups(); i++) {
+    candidates = lrcGroups.poiCandidates(i);
+  }
+  if (candidates.empty()) {
+    candidates = forwardingLrcGroups.nonPoiCandidates();
+    firstPoiCandidate = candidates.size();
+  }
+  for (size_t i = 0;
+       candidates.empty() && i < forwardingLrcGroups.numPoiGroups(); i++) {
+    candidates = forwardingLrcGroups.poiCandidates(i);
+    forwardingInfo = forwardingLrcForwardingTo[i];
   }
```

Comment:
You can achieve something like this with a "first non-empty group" method that just:

1. checks if non-poi groups has been set, and if it's set and non-empty, returns a pointer to that
2. iterates over the POI groups, and returns a pointer the first non-empty one
3. Returns a nullptr otherwise
4. Maybe an out-param `ssize_t` for "index of POI candidate group, if any?

Here, your code becomes:

```C++
  ssize_t poiGroup = -1;
  if (candidates.empty()) {
    if (auto lastResortCandidates = lrcGroups.firstNonEmptyGroup(poiGroup)) {
      candidates = *lastResortCandidates;
      if (poiGroup == -1) firstPoiCandidate = candidates.size();
    }
    if (auto lastResortCandidates = forwardingLrcGroups.firstNonEmptyGroup(poiGroup)) {
      candidates = *lastResortCandidates;
      if (poiGroup == -1) firstPoiCandidate = candidates.size();
      else forwardingInfo = forwardingLrcForwardingTo[poiGroup];
    }
  }
```

Then, some of that code looks like it can be factored out.

```C++
static void tryUseLastResortCandidates(const LastResortCandidateGroups& groups,
                                       CandidatesVec& outCandidates,
                                       size_t outFirstPoiCandidate,
                                       ForwardingInfoVec* outForwardngInfo = nullptr,
                                       std::vector<ForwardingInfoVec>* forwardingInfoPerPoi = nullptr) {
  ssize_t poiGroup = -1;
  if (outCandidates.empty()) {
    if (auto lastResortCandidates = groups.firstNonEmptyGroup(poiGroup)) {
      outCandidates = *lastResortCandidates;
      if (poiGroup == -1) {
        outFirstPoiCandidate = outCandidates.size();
      } else if (outForwardngInfo && forwardingInfoPerPoi) {
        *outForwardngInfo = (*forwardingInfoPerPoi)[poiGroup];
    }
  }
}
```

Finally,

```C++
  tryUseLastResortCandidates(lrcGroups, candidates, firstPoiCandidate);
  tryUseLastResortCandidates(forwardingLrcGroups, candidates, firstPoiCandidate, &forwardingInfo, &forwardingLrcForwardingTo);
```

Just a thought. You don't have to follow the suggestion, of course.

Reply:
I like the "first non-empty group" idea and will implement that. In your first chunk of code I think the `if` that looks through `forwardingLrcGroups` needs to be an `else if`, since we don't want to consider forwarding candidates if we found a match in non-forwarding candidates.

I'll skip refactoring into a separate function, as once it's simplified down I personally prefer the plain repetition over the extra parameters and logic in the function for the forwarding case.

Reply:
With the refactors I made to handle forwarding properly it ended up making the most sense to put searching for forwarded-to candidates into the same firstNonEmpty method, so I ended up pretty much applying the suggestion after all.